contracts/adapters/swap/source/src/contract.rs

```diff
@@ -0,0 +1,389 @@
+    let expected_denom_out = operations.last().unwrap().denom_out.clone();
+
+    let coin = get_coin_from_asset(asset)?;
+
+    let router = ROUTER_CONTRACT_ADDRESS.load(deps.storage)?;
+
+    let query_msg = swap_source::QueryMsg::SimulateSwapExactAmountInWithHops {
+        input: coin,
+        hops: operations
+            .into_iter()
+            .map(|op| Hop {
+                pool: op.pool,
+                denom: op.denom_out,
+            })
+            .collect(),
+    };
+
+    let router_resp = deps
+        .querier
+        .query_wasm_smart::<swap_source::SwapExactAmountInWithHopsResponse>(router, &query_msg)?;
+
+    let coin_out = Coin {
+        denom: expected_denom_out,
+        amount: router_resp.coin_out,
+    };
+
+    Ok(SimulateSwapExactAssetInResponse {
+        asset_out: coin_out.into(),
+        spot_price: price.then_some(router_resp.spot_price),
+    })
+}
+
+fn query_simulate_smart_swap_exact_asset_in_with_metadata(
+    deps: Deps,
+    asset_in: Asset,
+    ask_denom: String,
+    routes: Vec<Route>,
+    include_price: bool,
+) -> Result<SimulateSmartSwapExactAssetInResponse, ContractError> {
+    let (asset_out, spot_price) =
+        simulate_smart_swap_exact_asset_in(deps, asset_in, ask_denom, &routes, include_price)?;
+
+    Ok(SimulateSmartSwapExactAssetInResponse {
+        asset_out,
+        spot_price,
+    })
+}
+
+fn query_simulate_smart_swap_exact_asset_in(
+    deps: Deps,
+    asset_in: Asset,
+    ask_denom: String,
+    routes: &[Route],
+) -> Result<Asset, ContractError> {
+    simulate_smart_swap_exact_asset_in(deps, asset_in, ask_denom, routes, false).map(|r| r.0)
+}
+
+fn simulate_smart_swap_exact_asset_in(
+    deps: Deps,
+    asset_in: Asset,
+    ask_denom: String,
+    routes: &[Route],
+    include_price: bool,
+) -> Result<(Asset, Option<Decimal>), ContractError> {
+    let mut asset_out = Asset::new(deps.api, &ask_denom, Uint128::zero());
+
+    let mut weighted_price = Decimal::zero();
+
+    for route in routes {
+        let swap_in_with_meta = query_simulate_swap_exact_asset_in_with_metadata(
+            deps,
+            route.offer_asset.clone(),
+            route.operations.clone(),
+            include_price,
+        )?;
+
+        asset_out.add(swap_in_with_meta.asset_out.amount())?;
+
+        if include_price {
+            let weight = Decimal::from_ratio(route.offer_asset.amount(), asset_in.amount());
+            // the spot price returned by the swap baby router is the
+            // price of the cumulative swap operations.
+            weighted_price += swap_in_with_meta.spot_price.unwrap() * weight;
+        }
+    }
+
+    Ok((asset_out, include_price.then_some(weighted_price)))
+}
+
+fn get_coin_from_asset(asset: Asset) -> Result<Coin, ContractError> {
+    match asset {
+        Asset::Native(coin) => Ok(coin),
+        Asset::Cw20(_) => Err(ContractError::AssetNotNative),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    #[test]
+    fn test() {}
+}
```

Comment:
Cam remove since tests are in different file

Reply:
ok